lib/tsc-watch.js

```diff
@@ -8,12 +8,175 @@ const { extractArgs } = require('./args-manager');
+    compilationStarted: new Command(() => onCompilationStarted, 'started'),
+    compilationComplete: new Command(() => onCompilationComplete, 'complete'),
+  }
+
+  currentState = { state: 'not_running' };
+  desiredState = { state: 'not_running' };
+  nextRunID = 1;
+
+  /**
+   * Start running the given set of commands.
+   * If commands are already running, these will be killed first. (Even if the already running commands are identical to the new ones.)
+   * @param {*} commands An array of `Command`s to run.
+   * @param {boolean} isTriggeredByClient If true, this event was triggered by the `tsc-watch` client API, rather than an actual compilation event. 
+   */
+  run({ commands, isTriggeredByClient })
+  {
+    this._setDesiredState({ state: 'running', commands, isTriggeredByClient, runID: this.nextRunID++ });
+  }
+
+  /**
+   * Kill all running commands in preparation for process shutdown.
+   */
+  async shutdown()
+  {
+    await this._kill();
+  }
+
+  /**
+   * Set the desired state of the CommandRunner.
+   * This indicates which commands it should be running, if any.
+   * It is safe to call this at any time without worrying about the current state of the commands.
+   * @param {*} desiredState The new desired state.
+   * Note: `{ state: 'killing' }` is not a valid desired state - to kill a running process, set the desired
+   * state to `{ state: 'not_running' }`, and the state machine will transition to `not_running` via `killing`.
+   */
+  _setDesiredState(desiredState)
+  {
+    this.desiredState = desiredState;
+    this._stateMachine();
+  }
+
+  _stateMachine()
+  {
+    if (this.desiredState.state === 'running')
+    {
+      if (this.currentState.state === 'running' && this.currentState.runID !== this.desiredState.runID)
+      {
+        // running -> running
+        // We're already running some commands, and now we want to run some new ones.
+        // Need to kill the old commands first.
+        this._kill();
+      }
+      else if (this.currentState.state === 'not_running')
+      {
+        // not_running -> running
+        // We're not currently running anything, and we want to run some commands.
+        // We can go ahead and run them.
+        this._run(this.desiredState);
+      }
+    }
+    else if (this.desiredState.state === 'not_running')
+    {
+      if (this.currentState.state === 'running')
+      {
+        // running -> not_running
+        // We're currently running commands, and we don't want to be.
+        // Need to kill the commands that are running.
+        this._kill();
+      }
+    }
+  }
+
+  /**
+   * Implements the transition from the `not_running` state to the `running` state.
+   * @param {*} state The data for the new state.
+   */
+  _run({ commands, isTriggeredByClient, runID })
+  {
+    this.currentState = { state: 'running', commands, isTriggeredByClient, runID };
+    for (const command of commands)
+    {
+      command.run({ emitSignal: true });
```

Comment:
Question: Should `emitSignal` always be true, or only when `isTriggeredByClient` is false?